### PR TITLE
[SYCL] Deprecate the old-style SYCL attributes according to the SYCL 2020 spec

### PR DESCRIPTION
**Pull request by AaronBallman**
This adds the following spellings: `[[sycl::reqd_work_group_size]]`, `[[sycl::reqd_sub_group_size]]`, `[[sycl::work_group_size_hint]]`, and `[[sycl::vec_type_hint]]`. These spellings are allowed to be used in SYCL 2017 mode (with a warning about using an extension).

This deprecates the following spellings in SYCL 2020 mode: `__attribute__((vec_type_hint))`, `__attribute__((reqd_work_group_size))`, `[[cl::reqd_work_group_size]]`, and `__attribute__((work_group_size_hint))`.

### CODE REVIEW

**Comment by AaronBallman** [comment]
I did *not* deprecate any of the `[[intel::whatever]]` spellings that are wholly covered by the `[[sycl::whatever]]` spellings because it wasn't clear to me the `intel` spelling should be deprecated. If we decide we want that, we should do it separately.

Also, there are at least two issues that I think I found:

* When you enable `-fsycl-is-device` or `-fsycl-is-host` but don't pass `-syclstd=`, then the SYCL version reported by langopts is `SYCL_None`, which seems wrong. I worked around it.
* It looks like `[[sycl::work_group_size_hint]]` doesn't correctly handle template arguments or merging correctly. I added a FIXME comment to the tests about it, but that should be handled in a follow-up.

I also took the opportunity to add some documentation for `[[sycl::work_group_size_hint]]`.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3076]: It turns out that we've read the spec somewhat wrong here. The attribute does not have default arguments and each form with different parameter arity is expected to behave as a distinct attribute (see https://github.com/KhronosGroup/SYCL-Docs/issues/142).

I think a follow-up patch should detangle this situation as it's not really being made worse by these changes.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:3168]: The changes here are also incorrect per (https://github.com/KhronosGroup/SYCL-Docs/issues/142). I think the best approach is for me to back these changes out so that all three arguments are required, and then when we fix the `reqd_work_group_size` problems, we can add the same functionality to the `work_group_size_hint` attribute.

WDYT?

Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3168]: >>I think the best approach is for me to back these changes out so that all three arguments are required, and then when we fix the reqd_work_group_size problems, we can add the same functionality to the work_group_size_hint attribute.

I agree with you, @AaronBallman. FYI, I will start working on refactoring reqd_work_group_size and max_work_group_size attributes after this PR gets merged.

File: clang/test/SemaSYCL/num_simd_work_items_device.cpp
Comment (smanna12) [inline clang/test/SemaSYCL/num_simd_work_items_device.cpp:65]: [[sycl::reqd_work_group_size(5, 3, 3)]]  - this is SYCL 2020 attributes, we are not passing -sycl-std=2020 in command line. does it mean that we are seeing default to 2020 here calling clang -cc1 -fsycl-is-device?

Comment (AaronBallman) [reply clang/test/SemaSYCL/num_simd_work_items_device.cpp:65]: No, but you might be raising a good question -- should these SYCL 2020 attributes be available in earlier modes? I had assumed yes, but perhaps that's a bad assumption. Should we be warning when using the SYCL 2020 spelling but not in SYCL 2020 mode?

Comment (smanna12) [reply clang/test/SemaSYCL/num_simd_work_items_device.cpp:65]: >>should these SYCL 2020 attributes be available in earlier modes?

Yes, they do.

>>Should we be warning when using the SYCL 2020 spelling but not in SYCL 2020 mode?

I think we should give a warning otherwise we will have no difference between SYCL 2020 spelling [[sycl::]] vs [[intel::]]

Comment (AaronBallman) [reply clang/test/SemaSYCL/num_simd_work_items_device.cpp:65]: What's the goal of the warning? I could imagine a compatibility warning similar to `-Wpre-c++17-compat` (e.g., https://godbolt.org/z/rdbsedrGv) that users opt into when they want to know they're using a feature in a newer SYCL mode but still want to be compatible with an older SYCL mode, but that strikes me as 1) lower priority because there likely aren't many people who care about 1.2.1 compatibility, and 2) something we should be doing consistently for all attributes and other features. I could also imagine a compatibility warning similar to `-Wc++17-extensions` (e.g., https://godbolt.org/z/3jPc53f5r) that is default on to let users know they're using an extension from a newer SYCL mode while in an older SYCL mode. This one strikes me as being a potentially higher priority, but still should be done consistently for all features.

Comment (AaronBallman) [reply clang/test/SemaSYCL/num_simd_work_items_device.cpp:65]: Having thought about this a bit more, I think we should have the warnings about compatibility. I'll add those here, but we should open an issue to make sure we do this consistently for other features we've already worked on (if there are any, I'm not certain).

Comment (smanna12) [reply clang/test/SemaSYCL/num_simd_work_items_device.cpp:65]: >> but we should open an issue to make sure we do this consistently for other features we've already worked on

I agree

Comment (AaronBallman) [reply clang/test/SemaSYCL/num_simd_work_items_device.cpp:65]: Waiting on landing https://github.com/intel/llvm/pull/3702 before I do the additional warning work, as that sets the SYCL version to something sensible when it's not explicitly passed on cc1.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (premanandrao) [inline clang/lib/Sema/SemaDeclAttr.cpp:3096]:
```suggestion
  // [[intel::max_work_group_size]] all require exactly three arguments.
```

File: clang/test/SemaSYCL/reqd-work-group-size-device.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/reqd-work-group-size-device.cpp:27]: Could you change one of these to '(4, 1)' just so that variant is also checked?

**Review by premanandrao — COMMENTED** [review]
Just nits.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3096]: Good catch!

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:331]: I'm rewriting the comment entirely for clarity.

File: clang/test/SemaSYCL/reqd-work-group-size-device.cpp
Comment (AaronBallman) [reply clang/test/SemaSYCL/reqd-work-group-size-device.cpp:35]: Yes, but there's a new, exciting bug here that needs to be fixed (not a bug caused by my changes, just a misunderstanding of the spec that these tests highlight).

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [inline clang/lib/Sema/SemaDeclAttr.cpp:339]: Mention this as well please. I assume this and the cl namespace deprecation is part of SYCL2020 spec as mentioned in commit title but I think its nice having the details in commit message to make it easier for people not familiar with spec

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:3076]: Please open a bug report (to modify the attribute handling) for this unless one already exists. We've had a lot of issues with these set of attributes and I'm worried we're losing information between reviews

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [inline clang/lib/Sema/SemaDeclAttr.cpp:3101]: Can't we just simplify this to check attributes have 3 arguments, since we set the default values for 'optional' arguments prior to this?

File: clang/include/clang/Basic/Attr.td
Comment (elizabethandrews) [inline clang/include/clang/Basic/Attr.td:1454]: Can you file an issue to look into whether `[intel::whatever]` should be deprecated?

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [inline clang/lib/Sema/SemaDeclAttr.cpp:3120]: Do we not need this check for `WorkGroupSizeHint`

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:3101]: Or are you keeping it this way because we'll be removing the 'default values' since we have now discovered this is a bug?

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:3120]: Ok. I see there are tests checking this. Could you tell me where it is being enforced?

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3076]: Github Issues: https://github.com/intel/llvm/issues/3743

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [inline clang/lib/Sema/SemaDeclAttr.cpp:3160]: I think you might need similar change like this so that the attribute will accept one, two, and three arg form.

  for (unsigned i = 0; i < 3; ++i) {
    if (i < AL.getNumArgs() &&
       !checkUInt32Argument(S, AL, AL.getArgAsExpr(i), WGSize[i], i,
                              /*StrictlyUnsigned=*/true))
       return;

// FIXME: the attribute is like reqd_work_group_size in that it has a one, two,
// and three arg form that needs to be supported.
[[sycl::work_group_size_hint(1)]] void f4(); // expected-error {{'work_group_size_hint' attribute requires exactly 3 arguments}}
[[sycl::work_group_size_hint(1, 1)]] void f5(); // expected-error {{'work_group_size_hint' attribute requires exactly 3 arguments}}

File: clang/include/clang/Basic/AttrDocs.td
Comment (AaronBallman) [reply clang/include/clang/Basic/AttrDocs.td:2605]: Great catch!

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:334]: There was only one `[[cl::whatever]]` spelling and it was already listed. But the message was a bit inaccurate, so I updated it in other places.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:339]: That's also already mentioned (I listed the spellings we deprecated explicitly). LMK if you think there's more work for me to do in the commit message though!

File: clang/include/clang/Basic/Attr.td
Comment (AaronBallman) [reply clang/include/clang/Basic/Attr.td:1454]: I can if you'd like, but I think what makes sense there is to leave the `[[intel::whatever]]` spellings as undeprecated and available in whatever modes they make sense to support in, even if it's an identical spelling to another attribute. WDYT?

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3120]: This is handled automagically thanks to tablegen. The attribute lists three arguments, none of which are optional (https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/Attr.td#L2968) and so the common handler gets it (https://github.com/intel/llvm/blob/sycl/clang/lib/Sema/SemaAttr.cpp#L1244).

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3160]: I think we should handle the argument changes for `reqd_work_group_size` and `work_group_size_hint` at the same time, so for this patch, I've left the argument handling bits alone. Do you prefer I do those changes as part of this patch?

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:3101]: We can't simplify this check as it's checking the number of args passed to the parsed form of the attribute (so we've not applied any default values to it yet -- that happens when making the semantic attribute).

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:3160]: > I think we should handle the argument changes for `reqd_work_group_size` and `work_group_size_hint` at the same time, so for this patch, I've left the argument handling bits alone. Do you prefer I do those changes as part of this patch?

>>I think we should handle the argument changes for `reqd_work_group_size` and `work_group_size_hint` at the same time, so for this patch, I've left the argument handling bits alone. 

I agree. We should do this in follow-up patch (do not need to make any changes as part of this patch here). Thanks.

File: clang/include/clang/Basic/Attr.td
Comment (smanna12) [reply clang/include/clang/Basic/Attr.td:1454]: >>but I think what makes sense there is to leave the [[intel::whatever]] spellings as undeprecated and available in whatever modes they make sense to support in, even if it's an identical spelling to another attribute.

Yes, i agree.

FYI, CXX11<"intel", "sub_group_size">- this is new attribute that we added recently and this is part of CXX11<"intel", "reqd_sub_group_size">,

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:339]: Oh I see. I didn't realize we just had the explicitly mentioned spellings in those namespaces. My bad!

File: clang/test/SemaSYCL/reqd-sub-group-size-device.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/reqd-sub-group-size-device.cpp:22]: Is the 'expected-note' at the end of this for the conflict with the line below?

File: clang/test/SemaSYCL/sycl-2017-future-compat.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/sycl-2017-future-compat.cpp:9]: missing 'not'

File: clang/test/SemaSYCL/vec-type-hint.cpp
Comment (premanandrao) [inline clang/test/SemaSYCL/vec-type-hint.cpp:10]: And this not getting a warning in OpenCL mode is the behavior before this change as well?

File: clang/test/SemaSYCL/reqd-sub-group-size-device.cpp
Comment (AaronBallman) [reply clang/test/SemaSYCL/reqd-sub-group-size-device.cpp:22]: Correct. `The previous attribute is here` on the first declaration is from the second decl conflict, and the one on the second declaration is from the third decl conflict.

File: clang/test/SemaSYCL/sycl-2017-future-compat.cpp
Comment (AaronBallman) [reply clang/test/SemaSYCL/sycl-2017-future-compat.cpp:9]: Hehe, c'mon, maybe the test cases are what's wrong. Maybe we WANT diagnostics when using SYCL 2020 features in SYCL 2020 mode. Like "Congrats, you're using SYCL 2020. You win... uh... something, I guess." ;-)

File: clang/test/SemaSYCL/vec-type-hint.cpp
Comment (AaronBallman) [reply clang/test/SemaSYCL/vec-type-hint.cpp:10]: Correct!

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (elizabethandrews) [reply clang/lib/Sema/SemaDeclAttr.cpp:3120]: Thanks for explanation!

File: clang/test/SemaSYCL/reqd-work-group-size-device.cpp
Comment (elizabethandrews) [inline clang/test/SemaSYCL/reqd-work-group-size-device.cpp:37]: Sorry. I didn't catch this before cl*

File: clang/test/SemaSYCL/sycl-2017-future-compat.cpp
Comment (elizabethandrews) [inline clang/test/SemaSYCL/sycl-2017-future-compat.cpp:6]: I did not know we could verify diagnostics like this! That's pretty cool!

File: clang/test/SemaSYCL/vec-type-hint.cpp
Comment (elizabethandrews) [inline clang/test/SemaSYCL/vec-type-hint.cpp:8]: What is OpenCL mode here? Like when is it allowed? When we're not compiling SYCL code?

**Review by elizabethandrews — COMMENTED** [review]
LGTM. Just a couple of minor nits

File: clang/test/SemaSYCL/reqd-work-group-size-device.cpp
Comment (AaronBallman) [reply clang/test/SemaSYCL/reqd-work-group-size-device.cpp:37]: Good catch!

File: clang/test/SemaSYCL/vec-type-hint.cpp
Comment (AaronBallman) [reply clang/test/SemaSYCL/vec-type-hint.cpp:8]: I think this is a think-o on my part and the comment should be fixed. I mean to say that the GNU spelling is deprecated in SYCL 2020 mode without replacement, but is still allowed in SYCL 2017 and OpenCL mode.

**Review by elizabethandrews — COMMENTED** [review]
LGTM but you need to fix a couple of tests. I think @smanna12's patch fixing was num_simd_work_items_device was merged adding a few more cases in the tests.

**Comment by smanna12** [comment]
>  I think @smanna12's patch fixing was num_simd_work_items_device was merged adding a few more cases in the tests.

Yes, the test need to be updated with this current PR.

**Review by premanandrao — COMMENTED** [review]
LGTM modulo @elizabethandrews's comments.

**Comment by AaronBallman** [comment]
> LGTM but you need to fix a couple of tests. I think @smanna12's patch fixing was num_simd_work_items_device was merged adding a few more cases in the tests.

Thanks! Yeah, I missed that these were failing (locally too) from the merge. I've corrected the tests by keeping some of the old spellings and changing some to the new spellings. I tried to keep existing test coverage as close as possible, but if I accidentally lost some coverage with a change, feel free to mention it.

**Review by smanna12 — APPROVED** [review]
LGTM. Thank you

**Review by elizabethandrews — APPROVED** [review]
LGTM Thanks!

**Comment by AaronBallman** [comment]
The Lit_With_Cuda failures appear to be unrelated to my changes (it looks like a hung process).

**Pull request merged**
